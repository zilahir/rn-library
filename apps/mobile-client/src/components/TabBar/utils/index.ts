export { default as icons } from "./icons";

export { onPress } from "./navigations";
