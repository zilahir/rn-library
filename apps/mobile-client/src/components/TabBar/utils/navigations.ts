import { BottomTabBarProps } from "@react-navigation/bottom-tabs";

interface INavigationProps {
  navigation: BottomTabBarProps["navigation"];
  targetKey: string;
}

/**
 *
 * @param root0
 * @param root0.navigation
 * @param root0.target
 * @param root0.isFocused
 * @param root0.targetKey
 * @param root0.targetName
 */
export function onPress({
  navigation,
  targetKey,
  isFocused,
  targetName,
}: INavigationProps & { isFocused: boolean; targetName: string }): void {
  const event = navigation.emit({
    type: "tabPress",
    target: targetKey,
    canPreventDefault: true,
  });

  if (!isFocused && !event.defaultPrevented) {
    navigation.navigate(targetName);
  }
}

/*
export function onLongPress({ navigation, targetKey }: INavigationProps): void {
  navigation.emit({
    type: "tabLongPress",
    target: targetKey,
  });
}
*/
