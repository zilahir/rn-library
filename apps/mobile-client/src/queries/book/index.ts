import { useQuery, UseQueryResult } from "@tanstack/react-query";
import { Book } from "pikkukirjasto-types/types/book";

import bookQueryKeyFactory from "./queryKeyFactory"; // TODO: change this import
import { fetchAllBooks } from "./cloudFunctions"; // TODO: change this import

/**
 * @description React hook for fetching all books
 * @returns {UseQueryResult<Book[]>} React Query result object
 */
function useBooks(): UseQueryResult<Book[]> {
  const books = useQuery({
    queryKey: bookQueryKeyFactory.all,
    queryFn: fetchAllBooks,
    refetchOnWindowFocus: false,
    retry: false,
    initialData: [],
    keepPreviousData: true,
  });

  return books;
}

export default useBooks;
