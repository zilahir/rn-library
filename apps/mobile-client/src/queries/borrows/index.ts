import { useQuery, UseQueryResult } from "@tanstack/react-query";
import { Borrow } from "pikkukirjasto-types/types/borrow";
import useDeviceId from "@app/hooks/useDeviceId";

import fetchUserBorrows from "./cloudFunctions/fetchUserBorrows";
import borrowsQueryKeyFactory from "./queryKeyFactory";

/**
 * @description React hook for fetching all borrows
 * @returns {UseQueryResult<Borrow[]>} React Query result object
 * @example const { data: borrows } = useBorrows();
 */
function useBorrows(): UseQueryResult<Borrow[]> {
  const { data: deviceId = "" } = useDeviceId();
  const books = useQuery({
    queryKey: [borrowsQueryKeyFactory.getBorrosByUser(deviceId)],
    queryFn: async () => fetchUserBorrows(deviceId),
    refetchOnWindowFocus: false,
    retry: false,
    initialData: [],
    keepPreviousData: true,
  });

  return books;
}

export default useBorrows;
