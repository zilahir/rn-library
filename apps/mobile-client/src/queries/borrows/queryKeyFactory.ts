const borrowsQueryKeyFactory = {
  all: ["book"] as const,
  getBorrosByUser: (userId: string) =>
    [...borrowsQueryKeyFactory.all, userId] as const,
};

export default borrowsQueryKeyFactory;
