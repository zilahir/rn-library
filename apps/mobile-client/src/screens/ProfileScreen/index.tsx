import { ReactElement } from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { useSharedValue } from "react-native-reanimated";
import { useTheme } from "@shopify/restyle";
import { Feather } from "@expo/vector-icons";
import { useTranslation } from "react-i18next";
import { FlatList } from "react-native-gesture-handler";
import withSafeAreaView from "@app/components/WithSafeAreaView";
import MainLayout from "@app/layouts/MainLayout";
import { Theme } from "@app/theme";
import useBorrows from "@app/queries/borrows";
import Loader from "@app/components/common/Loader";

import BookItem from "../HomeScreen/components/BookItem";

const styles = ({
  spacing,
  colors,
  borderRadii,
}: {
  spacing: Theme["spacing"];
  colors: Theme["colors"];
  borderRadii: Theme["borderRadii"];
}): ReturnType<typeof StyleSheet.create> =>
  StyleSheet.create({
    profileContainer: {
      paddingVertical: spacing.s,
      flexDirection: "row",
      alignItems: "center",
      marginBottom: spacing.m,
    },
    profileIcon: {
      borderRadius: 100,
      borderWidth: 2,
      borderColor: colors.orange,
      marginRight: spacing.m,
      padding: spacing.s,
    },
    welcomeText: {
      fontSize: 22,
      fontWeight: "500",
    },
    borrowContainer: {
      flexDirection: "row",
    },
    borrowedBooksText: {
      fontSize: 20,
      fontWeight: "600",
    },
    borrowedBooksDetailsText: {
      fontSize: 13,
      color: colors.gray,
    },
    textContainer: {
      flex: 1,
    },
    seeAllBorrowedBooksButtonContainer: {
      width: 100,
    },
    borrowedBooksContainer: {
      justifyContent: "center",
      marginVertical: spacing.l,
    },
    userSettingsContainer: {},
    shadow: {
      shadowColor: "#000",
      shadowOffset: {
        width: 0,
        height: 1,
      },
      shadowOpacity: 0.05,
      shadowRadius: 3.84,
      elevation: 5,
    },
    settingsSection: {
      backgroundColor: colors.white,
      borderRadius: borderRadii.l,
      paddingHorizontal: spacing.m,
      paddingVertical: spacing.s,
      marginVertical: spacing.m,
    },
    settingsItem: {
      flexDirection: "row",
      alignItems: "center",
      height: 50,
      gap: spacing.s,
    },
    settingsText: {
      fontWeight: "500",
      fontSize: 18,
    },
    loaderContainer: {
      justifyContent: "center",
      alignItems: "center",
    },
  });

/**
 *
 * @returns {ReactElement} - react component
 * @description - renders the profile screen
 */
function ProfileScreen(): ReactElement {
  const { spacing, colors, borderRadii } = useTheme<Theme>();
  const { t } = useTranslation();

  const animationValue = useSharedValue(0);

  const style = styles({ spacing, colors, borderRadii });

  const { data: books = [], isLoading } = useBorrows();

  return (
    <MainLayout>
      <View style={[style.profileContainer]}>
        <View style={[style.profileIcon]}>
          <Feather name="user" size={24} color={colors.orange} />
        </View>
        <Text style={[style.welcomeText]}>{t("profileScreen.welcome")}</Text>
      </View>
      <View style={[style.borrowContainer]}>
        <View style={[style.textContainer]}>
          <Text style={[style.borrowedBooksText]}>
            {t("profileScreen.borrowedBooks")}
          </Text>
          <Text style={[style.borrowedBooksDetailsText]}>
            {t("profileScreen.borrowedBooksDetails", { count: books.length })}
          </Text>
        </View>
      </View>
      <View style={[style.borrowedBooksContainer]}>
        {isLoading && (
          <View style={[style.loaderContainer]}>
            <Loader />
          </View>
        )}
        {!isLoading && books.length > 0 && (
          <FlatList
            testID="borrowed-books-list"
            data={books}
            showsHorizontalScrollIndicator={false}
            horizontal
            renderItem={({ item: borrow }): ReactElement => (
              <BookItem
                animationValue={animationValue}
                noShadow
                bookCardvariant="bookCardHorizontal"
                item={{
                  ...borrow.book,
                }}
              />
            )}
          />
        )}
      </View>
      <View style={[style.userSettingsContainer]}>
        <Text style={[style.borrowedBooksText]}>
          {t("profileScreen.settings")}
        </Text>
        <View style={[style.settingsSection, style.shadow]}>
          <TouchableOpacity style={[style.settingsItem]}>
            <Feather name="user" size={24} color={colors.orange} />
            <Text style={[style.settingsText]}>
              {t("profileScreen.account")}
            </Text>
          </TouchableOpacity>

          <TouchableOpacity style={[style.settingsItem]}>
            <Feather name="shield" size={24} color={colors.orange} />
            <Text style={[style.settingsText]}>
              {t("profileScreen.account")}
            </Text>
          </TouchableOpacity>
        </View>
        <View style={[style.settingsSection, style.shadow]}>
          <TouchableOpacity style={[style.settingsItem]}>
            <Feather name="log-out" size={24} color={colors.orange} />
            <Text style={[style.settingsText]}>
              {t("profileScreen.logout")}
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    </MainLayout>
  );
}

export default withSafeAreaView(ProfileScreen);
