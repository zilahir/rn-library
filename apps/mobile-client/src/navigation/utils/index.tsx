import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";
import HomeScreen from "@app/screens/HomeScreen";
import ProfileScreen from "@app/screens/ProfileScreen";
import TrendingScreen from "@app/screens/TrendingScreen";
import LibraryScreen from "@app/screens/LibraryScreen";

type ScreenName = keyof BottomTabParamList;

interface BottomTabScreen {
  label?: ScreenName;
  name: ScreenName;
  tabComponent: ComponentProps<any>;
  hideLabel?: boolean;
  isFloating?: boolean;
}

interface IBottomTabScreens {
  screens: BottomTabScreen[];
  getBottomTabBarTabs: () => BottomTabScreen[];
  getNumberOfTabBarTabs: () => number;
}

export const bottomTabScreens: IBottomTabScreens = {
  screens: [
    {
      label: "Home",
      name: "Home",
      tabComponent: HomeScreen,
    },

    {
      label: "Profile",
      name: "Profile",
      tabComponent: ProfileScreen,
    },
    {
      label: "Trending",
      name: "Trending",
      tabComponent: TrendingScreen,
    },
    {
      label: "Library",
      name: "Library",
      tabComponent: LibraryScreen,
    },
  ],
  getBottomTabBarTabs: (): BottomTabScreen[] =>
    bottomTabScreens.screens.map((screen) => ({
      ...screen,
      hideLabel: !!screen.hideLabel,
      isFloating: !!screen.isFloating,
    })),

  getNumberOfTabBarTabs: (): number => bottomTabScreens.screens.length,
};
