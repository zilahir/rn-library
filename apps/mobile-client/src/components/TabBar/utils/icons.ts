import { Feather } from "@expo/vector-icons";
import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";

const icons: Record<
  keyof BottomTabParamList,
  ComponentProps<typeof Feather>["name"]
> = {
  Home: "home",
  Profile: "user",
  Trending: "trending-up",
  Library: "book-open",
  Borrow: "book",
};

export default icons;
