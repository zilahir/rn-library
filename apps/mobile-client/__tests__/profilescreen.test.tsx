// @ts-expect-error this is a mock
import mockRNDeviceInfo from "react-native-device-info/jest/react-native-device-info-mock";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import ProfileScreen from "@app/screens/ProfileScreen";
import * as hooks from "@app/queries/borrows";

import { render } from "./utils/wrapper";

const demoBooks = [
  {
    isbn: "demo isbn",
    author: "Terrell Stroman",
    title: "dedico adhuc tripudio",
    cover: "https://picsum.photos/seed/9YjBfGro/640/480",
  },
];

jest.mock("react-native/Libraries/EventEmitter/NativeEventEmitter");
jest.mock("react-native-device-info", () => mockRNDeviceInfo);

describe("ProfileScreen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("reenders correctly", () => {
    const { toJSON } = render(
      <QueryClientProvider client={new QueryClient()}>
        <ProfileScreen />
      </QueryClientProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("should not present a loader", () => {
    const { queryByTestId } = render(
      <QueryClientProvider client={new QueryClient()}>
        <ProfileScreen />
      </QueryClientProvider>
    );

    const loader = queryByTestId("loader");
    expect(loader).toBeNull();
  });

  it("should present a loader", () => {
    jest.spyOn(hooks, "default").mockImplementation((): any => ({
      data: [...demoBooks],
      isLoading: true,
    }));

    const { getByTestId } = render(
      <QueryClientProvider client={new QueryClient()}>
        <ProfileScreen />
      </QueryClientProvider>
    );

    const loader = getByTestId("loader");
    expect(loader).toBeDefined();
  });

  it("booklist should be present", () => {
    jest.spyOn(hooks, "default").mockImplementation((): any => ({
      data: [...demoBooks],
      isLoading: false,
    }));

    const { getByTestId } = render(
      <QueryClientProvider client={new QueryClient()}>
        <ProfileScreen />
      </QueryClientProvider>
    );

    const booklist = getByTestId("borrowed-books-list");
    expect(booklist).toBeDefined();
  });
});
