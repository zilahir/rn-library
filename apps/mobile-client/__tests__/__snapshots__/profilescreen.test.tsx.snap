// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileScreen reenders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderBottomLeftRadius": 25,
        "borderBottomRightRadius": 25,
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
          "marginHorizontal": 24,
          "marginVertical": 16,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "paddingVertical": 8,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "borderColor": "#FF9900",
              "borderRadius": 100,
              "borderWidth": 2,
              "marginRight": 16,
              "padding": 8,
            },
          ]
        }
      >
        <Text />
      </View>
      <Text
        style={
          [
            {
              "fontSize": 22,
              "fontWeight": "500",
            },
          ]
        }
      >
        profileScreen.welcome
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flex": 1,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 20,
                "fontWeight": "600",
              },
            ]
          }
        >
          profileScreen.borrowedBooks
        </Text>
        <Text
          style={
            [
              {
                "color": "#828282",
                "fontSize": 13,
              },
            ]
          }
        >
          profileScreen.borrowedBooksDetails
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "justifyContent": "center",
            "marginVertical": 24,
          },
        ]
      }
    />
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 20,
              "fontWeight": "600",
            },
          ]
        }
      >
        profileScreen.settings
      </Text>
      <View
        style={
          [
            {
              "backgroundColor": "#fff",
              "borderRadius": 25,
              "marginVertical": 16,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            {
              "elevation": 5,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.05,
              "shadowRadius": 3.84,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 8,
              "height": 50,
              "opacity": 1,
            }
          }
        >
          <Text />
          <Text
            style={
              [
                {
                  "fontSize": 18,
                  "fontWeight": "500",
                },
              ]
            }
          >
            profileScreen.account
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 8,
              "height": 50,
              "opacity": 1,
            }
          }
        >
          <Text />
          <Text
            style={
              [
                {
                  "fontSize": 18,
                  "fontWeight": "500",
                },
              ]
            }
          >
            profileScreen.account
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "#fff",
              "borderRadius": 25,
              "marginVertical": 16,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            {
              "elevation": 5,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.05,
              "shadowRadius": 3.84,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 8,
              "height": 50,
              "opacity": 1,
            }
          }
        >
          <Text />
          <Text
            style={
              [
                {
                  "fontSize": 18,
                  "fontWeight": "500",
                },
              ]
            }
          >
            profileScreen.logout
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
